File: tests/test_cli_project_commands.py
#!/usr/bin/env python3
"""
tests/test_cli_project_commands.py — Tests for CLI 'konoha project' and 'konoha data --project' commands.
Verifies:
1. konoha project context [path]
2. konoha project list
3. konoha project add [path] "<summary>"
4. konoha project memory [path]
5. konoha project delete <path|hash>
6. konoha data memory [agent] --project [path]
"""

import sys
import os
import tempfile
import unittest
import json
import subprocess
from pathlib import Path

REPO_ROOT = Path(__file__).parent.parent.resolve()
CLI_JS = REPO_ROOT / "bin" / "cli.js"

import re

def strip_ansi(text):
    return re.sub(r'\x1b\[[0-9;]*[mGKF]', '', text)

class TestCLIProjectCommands(unittest.TestCase):
    def setUp(self):
        self.tmp_dir = tempfile.TemporaryDirectory()
        self.project_dir = os.path.join(self.tmp_dir.name, "ecommerce_test_repo")
        os.makedirs(self.project_dir, exist_ok=True)

        with open(os.path.join(self.project_dir, "package.json"), "w") as f:
            json.dump({
                "name": "my-ecommerce-project",
                "dependencies": {
                    "next": "16.0.0",
                    "tailwindcss": "^4.0.0"
                }
            }, f)

    def tearDown(self):
        self.tmp_dir.cleanup()

    def run_cli(self, args):
        cmd = ["node", str(CLI_JS)] + args
        res = subprocess.run(cmd, capture_output=True, text=True, cwd=str(REPO_ROOT))
        return res

    def test_project_context_command(self):
        res = self.run_cli(["project", "context", self.project_dir])
        self.assertEqual(res.returncode, 0, f"Error: {res.stderr}")
        self.assertIn("my-ecommerce-project", strip_ansi(res.stdout))
        self.assertIn("Next.js", strip_ansi(res.stdout))
        self.assertIn("pnpm", strip_ansi(res.stdout))

    def test_project_add_and_list_command(self):
        # 1. Add invariant
        res_add = self.run_cli(["project", "add", self.project_dir, "Always enforce strict typing and dark mode theme variables"])
        self.assertEqual(res_add.returncode, 0)
        self.assertIn("Saved architectural invariants", strip_ansi(res_add.stdout))

        # 2. List projects
        res_list = self.run_cli(["project", "list"])
        self.assertEqual(res_list.returncode, 0)
        self.assertIn("my-ecommerce-project", strip_ansi(res_list.stdout))

    def test_project_memory_and_delete(self):
        # 1. Add context first
        self.run_cli(["project", "add", self.project_dir, "Test invariant for memory check"])

        # 2. View project memory
        res_mem = self.run_cli(["project", "memory", self.project_dir])
        self.assertEqual(res_mem.returncode, 0)

        # 3. Delete project profile
        res_del = self.run_cli(["project", "delete", self.project_dir])
        self.assertEqual(res_del.returncode, 0)
        self.assertIn("Deleted project profile", strip_ansi(res_del.stdout))

if __name__ == '__main__':
    unittest.main()

File: tests/test_cli_project_commands.js
#!/usr/bin/env node
'use strict';

/**
 * tests/test_cli_project_commands.js — Tests for CLI 'konoha project' and 'konoha data --project' commands.
 */

const assert = require('assert');
const fs = require('fs');
const os = require('os');
const path = require('path');
const { execFileSync } = require('child_process');

const REPO_ROOT = path.resolve(__dirname, '..');
const CLI_JS = path.join(REPO_ROOT, 'bin', 'cli.js');

function stripAnsi(text) {
  return text.replace(/\x1b\[[0-9;]*[mGKF]/g, '');
}

let activeTestDbPath = null;

function runCli(args) {
  try {
    const env = { ...process.env };
    if (activeTestDbPath) env.KONOHA_DB_PATH = activeTestDbPath;
    const stdout = execFileSync('node', [CLI_JS, ...args], {
      cwd: REPO_ROOT,
      encoding: 'utf-8',
      env,
      stdio: ['ignore', 'pipe', 'pipe'],
      timeout: 30000
    });
    return { status: 0, stdout, stderr: '' };
  } catch (err) {
    return {
      status: err.status || 1,
      stdout: err.stdout ? err.stdout.toString() : '',
      stderr: err.stderr ? err.stderr.toString() : err.message
    };
  }
}

async function run() {
  console.log('Running test_cli_project_commands tests...');
  const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), 'cli_proj_'));
  const testDbPath = path.join(tmpDir, 'test_konoha.db');
  activeTestDbPath = testDbPath;
  const personaMemory = require('../src/persona_memory');
  const db = require('../src/db');
  personaMemory.DB_PATH = testDbPath;
  db.DB_PATH = testDbPath;
  const conn = db.getConnection(testDbPath);
  db.setupSchema(conn);
  conn.close();

  const projectDir = path.join(tmpDir, 'ecommerce_test_repo');
  fs.mkdirSync(projectDir, { recursive: true });

  fs.writeFileSync(path.join(projectDir, 'package.json'), JSON.stringify({
    name: 'my-ecommerce-project',
    dependencies: {
      next: '16.0.0',
      tailwindcss: '^4.0.0'
    }
  }));

  try {
    // 1. project context
    const resCtx = runCli(['project', 'context', projectDir]);
    assert.strictEqual(resCtx.status, 0, `Error: ${resCtx.stderr}`);
    const cleanCtx = stripAnsi(resCtx.stdout);
    assert.ok(cleanCtx.includes('my-ecommerce-project'));
    assert.ok(cleanCtx.includes('Next.js'));
    assert.ok(cleanCtx.includes('pnpm'));
    console.log('✓ Project context command passed');

    // 2. project add and list
    const resAdd = runCli(['project', 'add', projectDir, 'Always enforce strict typing and dark mode theme variables']);
    assert.strictEqual(resAdd.status, 0, `Error: ${resAdd.stderr}`);
    assert.ok(stripAnsi(resAdd.stdout).includes('Saved architectural invariants'));

    const resList = runCli(['project', 'list']);
    assert.strictEqual(resList.status, 0, `Error: ${resList.stderr}`);
    assert.ok(stripAnsi(resList.stdout).includes('my-ecommerce-project'));
    console.log('✓ Project add and list command passed');

    // 3. project memory, memory delete, and project prune
    const personaMemory = require('../src/persona_memory');
    const memId = personaMemory.saveMemory({
      agentName: 'jonin',
      content: 'CLI Test Project Learning',
      projectPath: projectDir,
      dbPath: testDbPath
    });
    assert.ok(memId, 'Saved memory should return ID');

    const resMem = runCli(['project', 'memory', projectDir]);
    assert.strictEqual(resMem.status, 0, `Error: ${resMem.stderr}`);
    assert.ok(stripAnsi(resMem.stdout).includes('CLI Test Project Learning'));

    // Test delete-memory 1 by 1
    const resDelMem = runCli(['project', 'delete-memory', String(memId)]);
    assert.strictEqual(resDelMem.status, 0, `Error: ${resDelMem.stderr}`);
    assert.ok(stripAnsi(resDelMem.stdout).includes(`Deleted memory item ID: ${memId}`));

    // Add another memory and test prune
    personaMemory.saveMemory({
      agentName: 'anbu',
      content: 'Prune Test Learning',
      projectPath: projectDir,
      dbPath: testDbPath
    });

    const resPrune = runCli(['project', 'prune', projectDir, '--invariants']);
    assert.strictEqual(resPrune.status, 0, `Error: ${resPrune.stderr}`);
    assert.ok(stripAnsi(resPrune.stdout).includes('Pruned'));
    assert.ok(stripAnsi(resPrune.stdout).includes('Architectural invariants cleared'));

    // 4. project delete and prune-all
    const resDel = runCli(['project', 'delete', projectDir]);
    assert.strictEqual(resDel.status, 0, `Error: ${resDel.stderr}`);
    assert.ok(stripAnsi(resDel.stdout).includes('Deleted project profile'));

    const resPruneAll = runCli(['project', 'prune-all']);
    assert.strictEqual(resPruneAll.status, 0, `Error: ${resPruneAll.stderr}`);
    assert.ok(stripAnsi(resPruneAll.stdout).includes('Pruned all registered workspace profiles'));
    console.log('✓ Project memory, delete-memory, prune, and delete commands passed');

    console.log('\nAll test_cli_project_commands tests passed!');
  } finally {
    const personaMemory = require('../src/persona_memory');
    const db = require('../src/db');
    activeTestDbPath = null;
    personaMemory.DB_PATH = null;
    db.DB_PATH = null;
    try {
      fs.rmSync(tmpDir, { recursive: true, force: true });
    } catch (_) {}
  }
}

if (require.main === module) {
  run().catch(err => {
    console.error(err);
    process.exit(1);
  });
}

module.exports = { run };
